src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/ProjectInfoEndpoint.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.Serialization;
+using Microsoft.AspNetCore.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+
+/// <summary>
+/// Used to receive project system info updates from the client that were discovered OOB.
+/// </summary>
+[RazorLanguageServerEndpoint(LanguageServerConstants.RazorProjectInfoEndpoint)]
+internal class ProjectInfoEndpoint : IRazorNotificationHandler<ProjectInfoParams>, IDisposable
+{
+    private readonly ProjectConfigurationStateManager _projectConfigurationStateManager;
+
+    public ProjectInfoEndpoint(ProjectConfigurationStateManager projectConfigurationStateManager)
+    {
+        _projectConfigurationStateManager = projectConfigurationStateManager;
+    }
+
+    public bool MutatesSolutionState => false;
```

Comment:
Related to my comments about whether we want to delay updates, we might want this to be true

Reply:
You mean it should be true if we don't batch, right?

Reply:
Yeah. Setting to true means the server will block processing any LSP request that comes in, until this method is finished, which _might_ be a good thing if we're not batching. If we're batching then its not going to matter either way. This is another one that will be hard to measure though. Are we causing issues by holding everything else up? Are we seeing odd behaviour if we don't? Probably something to leave for the future, the more I think about it.

Reply:
I'm leaving it as false for now pending final testing. Let me know if you thing it should be true with the current code and what the likely scenario would be that would get us in trouble without it :)